docs/proposals/20190708-MachineConfig.md

```diff
@@ -0,0 +1,204 @@
+          node-labels: ""
+      ---
+      apiVersion: kubeadm.k8s.io/v1beta1
+      kind: ClusterConfiguration
+      clusterName: {{ .ClusterName }}
+      kubernetesVersion: {{ .KubernetesVersion }}
+      controlPlaneEndpoint: {{ .IP }}:443
+      apiServer:
+        certSANs: [ {{ range $i,$ip := .MasterIPs }}{{if $i}},{{end}}"{{$ip}}"{{end}}, "127.0.0.1" ]
+        extraArgs:
+          runtime-config: settings.k8s.io/v1alpha1=true
+          feature-gates: ExperimentalCriticalPodAnnotation=true
+      controllerManager:
+        extraArgs:
+          terminated-pod-gc-threshold: '100'
+          feature-gates: ExperimentalCriticalPodAnnotation=true
+      scheduler:
+        extraArgs:
+          feature-gates: ExperimentalCriticalPodAnnotation=true
+      networking:
+        dnsDomain: {{ .ServiceDomain }}
+        podSubnet: {{ index .PodNet 0 }}
+        serviceSubnet: {{ index .ServiceNet 0 }}
+      ---
+      apiVersion: kubelet.config.k8s.io/v1beta1
+      kind: KubeletConfiguration
+      featureGates:
+        ExperimentalCriticalPodAnnotation: true
+      ---
+      apiVersion: kubeproxy.config.k8s.io/v1alpha1
+      kind: KubeProxyConfiguration
+      mode: ipvs
+      ipvs:
+        scheduler: lc
+  trustd:
+    token: '{{ .TrustdInfo.Token }}'
+    endpoints: [ {{ .Endpoints }} ]
+    certSANs: [ "{{ .IP }}", "127.0.0.1" ]
+networking:
+  os:
+    devices:
+      - interface: lo
+        cidr: x.x.x.x/32
+      - interface: eth0
+        dhcp: true
+install:
+  wipe: false
+  force: true
+  boot:
+    device: /dev/sda
+    size: 1024000000
+    kernel: http://139.178.69.21:8080/assets/talos/release/v0.1.0-beta.0/vmlinuz
+    initramfs: http://139.178.69.21:8080/assets/talos/release/v0.1.0-beta.0/initramfs.xz
+  root:
+    device: /dev/sda
+    size: 2048000000
+    rootfs: http://139.178.69.21:8080/assets/talos/release/v0.1.0-beta.0/rootfs.tar.gz
+  data:
+    device: /dev/sda
+    size: 4096000000
+```
+
+## Proposal
+
+There are quite a few fields that are similar among all node types, while the main difference between them is the kubeadm templates. Given all the proper info, we should be able to generate the kubeadm configuration on the fly depending on node type and abstract the need to know all the different options away. I'm proposing a machine configuration that looks like this:
+
+```
+#!talos
+## We'll implement a versioning strategy and 
+## respect it when parsing the config
+version: "v0.1.0"
+## All security focused fields (certs and tokens for trustd/kubeadm)
+## go under the security key
+security:
+  os:
+    ca:
+      crt: "{{ .Certs.OsCert }}"
+      key: "{{ .Certs.OsKey }}"
+    token: abc.1234
+  kubernetes:
+    ca:
+      crt: "{{ .Certs.K8sCert }}"
+      key: "{{ .Certs.K8sKey }}"
+    token: abc.1234
+## All configs that are specific to this machine only
+## go in machine key
+machine:
+  nodeType: init, controlplane, worker
+  kubeletExtraArgs: {}
+  network:
+    interfaces:
```

Comment:
Updated with interfaces example provided by @Ulexus, as well as included the routes under the eth1 device as an example.